modelling/uncertainty_model/model.py

```diff
@@ -39,4 +175,127 @@ def no_feature_model(X):
+    model.compile(optimizer=opt, loss='mse')
+    # fit model
+    return model.fit([X, features], y, epochs=epochs, verbose=verbose, batch_size=batch_size)
+
+
+def get_conf_int_for(model, sample_id, X_test, y_test, features=None, num_samples=250, show_ci=True, print_ci=True,
+                     bins=25):
+    '''
+    Calculation of the 95% confidence interval for model predictions
+    :param model: keras model to test
+    :param sample_id: id to test
+    :param X_test: test sequences from PPG data
+    :param y_test: test targets
+    :param features: (optional) test features
+    :param num_samples: (optional) number of samples for the confidence interval
+    :param show_ci: (optional) plot prediction distribution
+    :param print_ci: (optional) print confidence interval, mean and true value
+    :param bins: (optional) number of bins for prediction distribution
+    :return: begin, end: confidence interval
+    '''
+    X_sample = np.tile(X_test[sample_id], (num_samples, 1, 1))
+    if features is not None:
+        features = np.tile(features[sample_id], (num_samples, 1))
+        ps = model.predict([X_sample, features])
+    else:
+        ps = model.predict(X_sample)
+
+    begin, end = sms.DescrStatsW(ps.flatten()).tconfint_mean()
+
+    if show_ci:
+        plt.hist(ps, bins=bins)
+        plt.show()
+    if print_ci:
+        print('Confidence interval: ', begin, end)
+        print('Mean: ', np.mean([begin, end]))
+        print('True: ', y_test[sample_id])
+        print('Classification: ', (begin <= y_test[sample_id]) and (y_test[sample_id] <= end))
+    return begin, end
+
+
+def get_conf_int_for_vectorized(model, X_test, y_test, features=None, num_samples=250, print_ci=True):
+    '''
+    Calculation of the 95% confidence interval for model predictions
+    :param model: keras model to test
+    :param sample_id: id to test
+    :param X_test: test sequences from PPG data
+    :param y_test: test targets
+    :param features: (optional) test features
+    :param num_samples: (optional) number of samples for the confidence interval
+    :param print_ci: (optional) print confidence interval, mean and true value
+    :return: begin, end: confidence interval
+    '''
+
+    X_sample = np.tile(X_test, (1, num_samples, 1)).reshape(-1, X_test.shape[1], X_test.shape[2])
+    fs = np.tile(features, (1, num_samples)).reshape(-1, features.shape[1])
+    ps = model.predict([X_sample, fs])
+
+    spl = np.split(ps, X_test.shape[0])
+
+    begins = []
+    ends = []
+    for i in range(len(spl)):
+        begin, end = sms.DescrStatsW(spl[i].flatten()).tconfint_mean()
+        begins.append(begin)
+        ends.append(end)
+
+    # if print_ci:
+    #     print('Confidence interval: ', begin, end)
+    #     print('Mean: ', np.mean([begin, end]))
+    #     print('True: ', y_test[sample_id])
+    #     print('Classification: ', (begin <= y_test[sample_id]) and (y_test[sample_id] <= end))
+
+    return np.array(begins), np.array(ends)
```

Comment:
Tried to vectorize the output distribution sampling, but it somehow ended up being even slower, so I just left it as is and didn't use it.